pkg/streamingpromql/operators/binary_operation.go

```diff
@@ -497,45 +547,180 @@ func (b *BinaryOperation) mergeOneSide(data []types.InstantVectorSeriesData, sou
 	}
 }
 
-func (b *BinaryOperation) computeResult(left types.InstantVectorSeriesData, right types.InstantVectorSeriesData) types.InstantVectorSeriesData {
-	var output []promql.FPoint
+func (b *BinaryOperation) mergeHistogramsOneSide(data []types.InstantVectorSeriesData, histogramOutput []promql.HPoint, sourceSeriesMetadata []types.SeriesMetadata, sourceSeriesIndices []int, side string) ([]promql.HPoint, error) {
+	remainingSeries := len(data)
+	for {
+		if remainingSeries == 1 {
+			// Only one series left, just copy remaining points.
+			for _, d := range data {
+				if len(d.Histograms) > 0 {
+					histogramOutput = append(histogramOutput, d.Histograms...)
+					return histogramOutput, nil
+				}
+			}
+		}
+
+		nextT := int64(math.MaxInt64)
+		sourceSeriesIndexInData := -1
+
+		for seriesIndexInData, d := range data {
+			if len(d.Histograms) == 0 {
+				continue
+			}
+
+			nextPointInSeries := d.Histograms[0]
+			if nextPointInSeries.T == nextT {
+				// Another series has a point with the same timestamp. We have a conflict.
+				firstConflictingSeriesLabels := sourceSeriesMetadata[sourceSeriesIndices[sourceSeriesIndexInData]].Labels
+				secondConflictingSeriesLabels := sourceSeriesMetadata[sourceSeriesIndices[seriesIndexInData]].Labels
+				groupLabels := b.labelsFunc()(firstConflictingSeriesLabels)
+
+				return nil, fmt.Errorf("found duplicate series for the match group %s on the %s side of the operation at timestamp %s: %s and %s", groupLabels, side, timestamp.Time(nextT).Format(time.RFC3339Nano), firstConflictingSeriesLabels, secondConflictingSeriesLabels)
+			}
+
+			if d.Histograms[0].T < nextT {
+				nextT = d.Histograms[0].T
+				sourceSeriesIndexInData = seriesIndexInData
+			}
+		}
+
+		histogramOutput = append(histogramOutput, data[sourceSeriesIndexInData].Histograms[0])
+		data[sourceSeriesIndexInData].Histograms = data[sourceSeriesIndexInData].Histograms[1:]
+
+		if len(data[sourceSeriesIndexInData].Histograms) == 0 {
+			remainingSeries--
+		}
+	}
+}
+
+func (b *BinaryOperation) computeResult(left types.InstantVectorSeriesData, right types.InstantVectorSeriesData) (types.InstantVectorSeriesData, error) {
+	var fPoints []promql.FPoint
+	var hPoints []promql.HPoint
 
 	// For one-to-one matching for arithmetic operators, reuse one of the input slices to avoid allocating another slice.
 	// We'll never produce more points than the smaller input side, so use that as our output slice.
 	//
 	// FIXME: this is not safe to do for one-to-many, many-to-one or many-to-many matching, as we may need the input series for later output series.
+	// FIXME: This is likely wrong when multiplying different types (eg histogram * float)
```

Comment:
Just had a thought on this: we could wait to allocate the slices until we have the first point of that type, similar to what we do for `sum`, rather than allocating them upfront like this